From bc4071aafcf4d0535ee423b69167696d6c03207d Mon Sep 17 00:00:00 2001
From: Evgeny Novikov <novikov@ispras.ru>
Date: Fri, 3 Jul 2020 14:15:18 +0300
Subject: [PATCH] hwmon: (aspeed-pwm-tacho) Avoid possible buffer overflow
Git-commit: bc4071aafcf4d0535ee423b69167696d6c03207d
Patch-mainline: v5.8-rc6
References: git-fixes

aspeed_create_fan() reads a pwm_port value using of_property_read_u32().
If pwm_port will be more than ARRAY_SIZE(pwm_port_params), there will be
a buffer overflow in
aspeed_create_pwm_port()->aspeed_set_pwm_port_enable(). The patch fixes
the potential buffer overflow.

Found by Linux Driver Verification project (linuxtesting.org).

Signed-off-by: Evgeny Novikov <novikov@ispras.ru>
Link: https://lore.kernel.org/r/20200703111518.9644-1-novikov@ispras.ru
Signed-off-by: Guenter Roeck <linux@roeck-us.net>
Acked-by: Takashi Iwai <tiwai@suse.de>

---
 drivers/hwmon/aspeed-pwm-tacho.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/drivers/hwmon/aspeed-pwm-tacho.c
+++ b/drivers/hwmon/aspeed-pwm-tacho.c
@@ -746,6 +746,8 @@ static int aspeed_create_fan(struct devi
 	ret = of_property_read_u32(child, "reg", &pwm_port);
 	if (ret)
 		return ret;
+	if (pwm_port >= ARRAY_SIZE(pwm_port_params))
+		return -EINVAL;
 	aspeed_create_pwm_port(priv, (u8)pwm_port);
 
 	count = of_property_count_u8_elems(child, "aspeed,fan-tach-ch");
